### ISSUE TITLE
Should we use a const generic for address validation instead of type?

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#3226]
One particularly annoying thing about `Address` is that derive macros often don't work properly with it. E.g. `#[derive(Clone)]` will put a bound on the type even though the type is not actually used. And this gets propagated if you have `NetworkValidation` downstream. Until recently I've assumed there's nothing we can do about it today. But I found out it's fixable.

If we change the type to `const NETWORK_VALIDATED: bool` then all derive macros will work correctly with it, we can get rid of annoying `PhantomData` and it might have other benefits related to not confusing the compiler into thinking that we store anything.

Disadvantages:
* Obviously, this is a breaking change, maybe we can workaround it by changing the validation enums to constants. Also maybe it won't blow up too much if people are mostly using `Address<_>` and `Address`.
* `Address<NetworkUnchecked>` is much more readable than `Address<false>` (sadly enums are not supported in const generics), conversely, people don't need to import annoying types
* It's possible that `parse::<Address<_>>()` will stop working, I'm not sure

It's also a question if there are other types where we can leverage constants for the same reason. We've discarded the idea of tracking whether hash preimage is known, so that's no longer relevant, for `Script` we would really need enums if we don't want to assign arbitrary numbers and have a bunch of constants.

**Comment by junderw** [comment]
> `Address<NetworkUnchecked>` is much more readable than `Address<false>` (sadly enums are not supported in const generics), conversely, people don't need to import annoying types

What about the following?

```rust
impl Address<B> {
    /// This is for readable use in the generic parameter
    pub const NETWORK_CHECKED: bool = true;
    /// This is for readable use in the generic parameter
    pub const NETWORK_UNCHECKED: bool = false;
}

fn foo() -> Address<Address::NETWORK_CHECKED> {
    todo!()
}

// or type alias

/// Network is unchecked
pub type AddressUnchecked = Address<false>;
/// Network is checked
pub type AddressChecked = Address<true>;
```

**Comment by Kixunil** [comment]
> are also less likely to lead to compile-time/size blowups

I'm skeptical about this since monomorphisation happens in both cases.

> deprecating the former because of its incorrect casing

Good idea

> But we should provide `NETWORK_CHECKED`/`UNCHECKED` constants and let them choose what they want to do.

Agreed

> And/or provide a new type alias `AddressUncheckedNetwork`.

That sounds also great. And we can improve it more: backport the type alias and deprecate `NetworkUncheceked` in favor of that type alias and `NetworkChecked` in favor of `Address`. Then if people do the fixes it won't break them at all unless they wrote `impl<T: NetworkValidation> Foo for Address<T> { ... }`

@junderw I like your idea of the constants being associated.

> It does suck that enums, even C-like enums, are not supported in const generics.

Yes, I'm pretty annoyed about this.

> _Although_ `char`s are, which could be a bit nicer than magic numbers at least.

Now this sounds interesting! It actually makes me want to do it. We should provide aliases/constants too of course. I propose these characters:

* 🔒 for script pubkey
* 🔑 for script sig
* 👁️ for witness v0 script
* 🥕 for TapScript

**Comment by tcharding** [comment]
> I propose these characters:

Best early morning chuckle I've had in weeks

**Comment by Kixunil** [comment]
@tcharding I'm actually pretty serious about it:

Benefits of emojis rather than chars like `s`, `p`...
* They represent a whole word which is more easily related to something, especially in case of 🥕 which was common symbol for Taproot on Twitter and other socials.
* They are annoying to write which incentivizes using constants instead of emojis and constants are even more readable
* They can be grepped. Try `rg p` and see why normal chars would be a problem
* They bring joy to programming. A little harmless joke here and there in the code can help people associate good feelings with the code.

Downsides:
* Some ancient editors/OSes can't show them. But if my `neovim` on Debian stable-based QubesOS template can do it why wouldn't anything else? (`apt install fonts-noto` is enough)
* Some regex engines count the non-ascii chars wrong (`....` is four bytes not chars for them), though that probably doesn't matter in a file that only defines constants.
* Some people might think we're nuts but a good documentation could fix that.

Seems that the benefits might outweigh the costs to me. And yes, I realize this is an instance of "It's funny but true." :)

We could also just test it for a bit and see if we get complains. Or we could generate the constants randomly in a build script to discourage people from relying on them (I think it won't affect build determinism because constants get compiled-out).

**Comment by apoelstra** [comment]
I also laughed out loud at the emojis.

I agree with your principled arguments but I think they're overridden by the characters being impossible for people to type.

**Comment by apoelstra** [comment]
>We could also just test it for a bit and see if we get complains. Or we could generate the constants randomly in a build script to discourage people from relying on them (I think it won't affect build determinism because constants get compil

This did cross my mind, as a way to make the constants "behave like types" in the sense that you can't manually construct or manipulate them. Using emojis for this purpose isn't the worst idea. But if they show up in error messages people will think we're screwing around. Vs if `true` and `false` show up in error messages they'll be just as confused but blame the nature of software.

**Comment by Kixunil** [comment]
> the characters being impossible for people to type

This is not a problem because they'd be only "typed" (copy-pasted from web, LOL) once: in the constants file and never touched again. You can copy it from the error message when you're `rg`ing. (I always copy types anyway.) I've also remembered there's a syntax for typing them as numbers (unicode). However this may be actually a good point:

> But if they show up in error messages people will think we're screwing around.

I was originally thinking that we would explain ourselves in the doc but if people see it in rustc output then they don't see it.

> you make a good point that ttys can't show them, many fonts are lacking them, uxrvt struggles with them

I thought the point is weaker since the development environment is usually setup to show these but I've just realized that is someone is compiling a broken branch of electrs on RPI then they'll not know how to report the bug seeing the placeholder. This may be actually a great counterargument.

Though, given all this I've just tested if type aliases are allowed in const generics. They are. Given this and ungreppability of single chars I have a different idea: we can pretty much emulate C-style enums by using `u64` with a unique number that's not used anywhere (either completely random or something like `u64::from_be_bytes(b"pubkey__")`) and we can declare the type behind the alias mildly unstable - if you use the cfg switch for inverse mutation testing then the type gets replaced by a different type with different constants. We could even use char with emojis here (written in the code as unicode codes to not mess up with tools too much). Then if in ten years they add C-style enums we can seamlessly switch to them. (I'd still probably do a major version bump but the upgrade will be painless.)

**Comment by Kixunil** [comment]
> because it's such an unidiomatic thing to do.

I don't think it's unidiomatic, it's juts that we're not used to it because `const` generics are new and people didn't have time to adapt.

A little backstory how I even realized this is a good idea: when we've first introduced it I was annoyed about having to put in `PhantomData` and breaking the API more than would be required otherwise (LOL, the irony) and I was thinking they should've allowed `T: 'static` to not have `PhantomData` because there's no issue around variance (maybe with an attribute to say "I didn't just forget to write it"). And yesterday I was thinking about it again and realized we do have it now in the form of const generics. (Actually, I kinda realized it before when doing some work on Matt's request to provide unbuffered API, I just didn't make the connection with `Address` until yesterday. BTW Most of that work was useless as there's a simpler solution but maybe it was still worth the experience.)

Anyway, OK, let's try it and revert it quickly if we find it too broken.

**Comment by apoelstra** [comment]
>I don't think it's unidiomatic, it's juts that we're not used to it because const generics are new and people didn't have time to adapt.

No, using `usize`s and magic values to emulate types **is** unidiomatic and nobody would ever consider doing it except that Rust's `derive` logic is broken. This is the kind of thing C programmers do to try to add some type safety to a language that lets you cast pointers to structures with no guardrails.

I'm not sure what kind of breakage `PhantomData` could cause. Either you have a POD type in which case adding marker types gains you nothing (since users can just bypass whatever the marker is supposed to represent), or you have a type that can only be constructed by function calls, in which case users don't care what private fields exist.

> Anyway, OK, let's try it and revert it quickly if we find it too broken.

Great, ACK.

**Comment by Kixunil** [comment]
> No, using `usize`s and magic values to emulate types **is** unidiomatic

`bool` field that uses both `true` and `fals` is idiomatic and you were talking about trying it in `addresses`. The constants are indeed weird. Actually one more issue with constants is that we cannot seal them, so people can define their own and they *will conflict* as opposed to types. For `Script` this is quite annoying. So yesterday I decided to drop my to-be-written proposal to allow `'static` tags but now I see I should write it anyway at some point.

> and nobody would ever consider doing it except that Rust's `derive` logic is broken

This is actually a bit of a rabbit hole. The broken logic looks obviously broken from the outside but there are good reasons why it is. `serde` fixed it with attributes allowing you to specify the bound but sadly while this was proposed in rustc it wasn't yet accepted. They'd prefer some nicer first-class approach, so we have to wait. (Oh, the downsides of stability.)

But I'd also argue if C-style enums were allowed in const generics this would be

> I'm not sure what kind of breakage `PhantomData` could cause

Previously we had:

```rust
struct Address {
    payload: Payload,
    network: Network,
}
```

Ignoring the fact that we later found that layout broken anyway, it forced us to add the `PhantomData` field which forced change of constructors for no good reason since that `PhantomData` shouldn't be needed if the tag is always `'static`.

> since users can just bypass whatever the marker is supposed to represent

That's good to talk about but in case of `Address` constructing is from within the code is always assumed to have correct network and not something that can be checked, so it was fine to keep it as POD (ignoring other things).

**Comment by apoelstra** [comment]
> That's good to talk about but in case of `Address` constructing is from within the code is always assumed to have correct network and not something that can be checked, so it was fine to keep it as POD (ignoring other things).

We took this strategy with Miniscript and it's made everything much harder for us, to have types which can be directly constructed and then might be invalid in all sorts of ways, can't have cached data attached to them, etc. I would recommend we try to just close any types that might have invariants.

(Though it's not always really possible; you could also argue that we should close `Transaction` and try to maintain invariants like a maximum serialized length or that there are always a nonzero number of inputs, and you'd quickly have an unusable type. I think having a POD type there is the right approach.)

**Comment by Kixunil** [comment]
> I would recommend we try to just close any types that might have invariants.

Definitely. The reason `Address` "was" special is that its invariant is related to code outside of `Address`. E.g. if my application was configured to run on regtest then the invariant is for `Address` to be compatible with regtest but to check it you need to supply the application configuration. Note that `Payload` still needs to be closed. However since we need different enums in different variants of `Payload` this entire comment is just academical. :)

> you could also argue that we should close `Transaction` and try to maintain invariants like a maximum serialized length or that there are always a nonzero number of inputs, and you'd quickly have an unusable type. I think having a POD type there is the right approach.

I think we kinda want both and will be able to at some point. The current `Transaction` is a dynamic low-level type with no invariants. Suitable mainly for constructing arbitrary transactions (if that's what your application is supposed to do). It should exist. But also other types representing a valid transaction (either entirely or with  some assumptions around previous ones or about the chain they are in) or even specific transaction templates ("always two outputs"). With the `AsTransaction` trait we'll be able to represent them. Some of them may find their way into the `bitcoin` crate, others (e.g. application-specific templates) will just live in external crates.

**Comment by jirijakes** [comment]
As the perpetrator of the original `NetworkValidation`, I would like to have a look and see what can be done. But I won't be too fast, so if somebody wants it very soon, please raise a hand.

**Comment by tcharding** [comment]
In case anyone else doesn't look too closely at the auto-mention links, there is case against this issue in https://github.com/rust-bitcoin/rust-bitcoin/pull/3256#issuecomment-2313646676

**Comment by Kixunil** [comment]
Oh, crap, I'm seriously thinking of canceling this proposal now. :( I thought it would be much less breaking.

**Comment by jirijakes** [comment]
Yeah. Additionally to breaking a lot, I believe the ergonomics of usage would be worse.
